ktor-io/common/test/io/ktor/utils/io/ByteChannelCloseTest.kt

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2014-2021 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.utils.io
+
+import kotlin.reflect.*
+import kotlin.test.*
+
+class ReadPacketWithoutExceptionByteChannelCloseTest : ByteChannelCloseTest(
+    ExpectedFailureException::class,
+    { close(ExpectedFailureException()) },
+    { readPacket(Int.MAX_VALUE) }
+)
+
+class ReadRemainingWithExceptionByteChannelCloseTest : ByteChannelCloseTest(
+    ExpectedFailureException::class,
+    { close(ExpectedFailureException()) },
+    { readRemaining() }
+)
+
+class ReadRemainingWithoutExceptionByteChannelCloseTest : ByteChannelCloseTest(
+    null,
+    { close() },
+    { readRemaining() }
+)
+
+abstract class ByteChannelCloseTest(
+    private val expectedExceptionClass: KClass<out Throwable>?,
+    private val closeChannel: ByteChannel.() -> Unit,
+    private val readFromChannel: suspend ByteChannel.() -> Unit
+) : ByteChannelTestBase() {
+
+    private inline fun assertMayFail(block: () -> Unit) {
+        when (expectedExceptionClass) {
+            null -> block()
+            else -> assertFailsWith(expectedExceptionClass, block)
+        }
+    }
+
+    @Test
+    fun testImmediateFailurePropagation() = runTest {
```

Comment:
This test and the ones below will run for every subclass. Maybe extract them to a separate class to avoid it?

Reply:
That's what I want :)
Those tests are a little modified tests from removed ReadRemainingTest.kt. Class contains several ways on how channel can be closed, and I thought, that it will be good base to check, that behavior of `readRemaining`/`readPacket` after channel closing (with or without exception) is more or less aligned.

Reply:
Oh, I see it now.